### PR TITLE
Add mysqld-exporter and node-exporter

### PR DESCRIPTION
**Pull request by zmraul**
# Issue

**Address [comment](https://github.com/canonical/mysql-operator/blob/d88132091903db71e898fe35c503852b2495567a/src/constants.py#L27) before merge**

Adresses [DPE-1255](https://warthogs.atlassian.net/browse/DPE-1255).

The charm needs to have some sort of monitoring built in. This allows an easier admin experience through the creation of dashboards, alerts...

# Solution
Add mysqld-exporter and node-exporter snaps to the charm. This will expose two endpoints to show metrics. The necessary libs have been added to integrate with COS bundle.

No need to review:
- `grafana_dashboard.py`
- `juju_topology.py`
- `snap.py`
- `prometheus_scrape.py`
- `src/grafana_dashboards/*`

# Testing
Once deployed, two endpoints can be accessed:
- `http://<ip-address>:9100/metrics`: node exporter metrics
- `http://<ip-address>:9104/metrics`: mysql exporter metrics

To integrate with COS, follow the documentation section on the PR.

# Release Notes

- `README.md`: Updated with monitoring section.
- `lib/---/mysql.py`: Now uses two more parameters (exporter user and password) and creates said user.
- `metadata.yaml`: Updated with required provides for COS integration.
- `src/`
  - `charm.py`: Added COS requirements, exporter password handling and mysqld-exporter configuration setting.
  - `constants.py`: Added exporter user related literals.
  - `mysql_vm_helpers.py`: Added snaps to install process and `connect_mysql_exporter` method to set snap configs.
- Unit tests have been updated and fixed to use the new parameters. 

[DPE-1255]: https://warthogs.atlassian.net/browse/DPE-1255?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/charms/mysql/v0/mysql.py [hunk 1]
```diff
@@ -276,6 +282,8 @@ def configure_mysql_users(self):
         configure_users_commands = (
             f"CREATE USER '{self.server_config_user}'@'%' IDENTIFIED BY '{self.server_config_password}'",
             f"GRANT ALL ON *.* TO '{self.server_config_user}'@'%' WITH GRANT OPTION",
+            f"CREATE USER '{self.exporter_user}'@'%' IDENTIFIED BY '{self.exporter_password}' WITH MAX_USER_CONNECTIONS 3",
+            f"GRANT PROCESS, REPLICATION CLIENT, SELECT ON *.* TO '{self.exporter_user}'@'%'",
```
Comment (paulomach) [inline lib/charms/mysql/v0/mysql.py:286]: A good starting point is [this](https://github.com/percona/percona-server-mysql-operator/blob/2684de79e4bbba6221a5c08a71c4799f622c4fd9/build/ps-entrypoint.sh#L318-L321) for the necessary `GRANTS`

File: lib/charms/mysql/v0/mysql.py [hunk 2]
```diff
@@ -276,6 +282,8 @@ def configure_mysql_users(self):
         configure_users_commands = (
```
Comment (paulomach) [inline lib/charms/mysql/v0/mysql.py:282]: @delgod , I had this talk already but I want to confirm with you if we are going to enable the `monitoring` user from the get go, as charm default, or if we should create it on demand when relating to the COS stack.

File: lib/charms/mysql/v0/mysql.py (hunk 1 above)
Comment (delgod) [reply lib/charms/mysql/v0/mysql.py:286]: yes, the link is correct, but without BACKUP_ADMIN
```suggestion
            f"CREATE USER '{self.exporter_user}'@'%' IDENTIFIED BY '{self.exporter_password}' WITH MAX_USER_CONNECTIONS 3",
            f"GRANT SYSTEM_USER, SELECT, PROCESS, SUPER, REPLICATION CLIENT, RELOAD ON *.* TO '{self.exporter_user}'@'%'",
```

File: lib/charms/mysql/v0/mysql.py (hunk 2 above)
Comment (delgod) [reply lib/charms/mysql/v0/mysql.py:282]: I want to have special/system users always

**Comment by zmraul** [comment]
> @zmraul , left some comments within the code. Besides that, how do you plan to control the exporter service?

At the moment the exporters can be on by default. If mysql is down for whatever reason, the exporter snap will just fail the scrape, and show that on the metrics endpoint.

File: README.md [hunk 3]
```diff
@@ -87,6 +87,41 @@ juju relate keystone:shared-db mysql:shared-db
...
+## Monitoring
+
+The Charmed MySQL Operator comes with several exporters by default. The metrics can be queried by accessing the following endpoints:
+
+- Node exporter: `http://<unit-ip>:9100/metrics`
+- MySQL exporter: `http://<unit-ip>:9104/metrics`
+
+Additionally, the charm provides integration with the [Canonical Observability Stack](https://charmhub.io/topics/canonical-observability-stack).
+
+Deploy cos-lite bundle in a Kubernetes environment. This can be done by following the [deployment tutorial](https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s). Since the Charmed MySQL Operator is deployed on a machine environment, it is needed to offer the endpoints of the COS relations. The [offers-overlay](https://github.com/canonical/cos-lite-bundle/blob/main/overlays/offers-overlay.yaml) can be used, and this step is shown on the COS tutorial.
```
Comment (shayancanonical) [inline README.md:99]: cos-lite -> COS-lite

File: lib/charms/mysql/v0/mysql.py [hunk 4]
```diff
@@ -232,6 +232,8 @@ def __init__(
         server_config_password: str,
         cluster_admin_user: str,
         cluster_admin_password: str,
+        exporter_user: str,
+        exporter_password: str,
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/mysql.py:236]: we will need to update the `LIBPATCH` of this file

File: src/constants.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 ROOT_USERNAME = "root"
 CLUSTER_ADMIN_USERNAME = "clusteradmin"
 SERVER_CONFIG_USERNAME = "serverconfig"
+EXPORTER_USERNAME = "exporter"
```
Comment (delgod) [inline src/constants.py:9]: Please use `monitoring`. a person who doesn't know about Prometheus should be able to understand for what user is needed from the name. 
```suggestion
EXPORTER_USERNAME = "monitoring"
```

File: src/mysql_vm_helpers.py [hunk 6]
```diff
@@ -163,6 +188,31 @@ def wait_until_mysql_connection(self) -> None:
...
+        Raises
+            snap.SnapError: if an issue occurs during config setting or restart
+        """
+        cache = snap.SnapCache()
+        mysqld_exporter = cache[MYSQL_EXPORTER_SNAP_NAME]
+
+        try:
+            mysqld_exporter.set(
+                {
+                    "mysql.host": self.instance_address,
```
Comment (delgod) [inline src/mysql_vm_helpers.py:203]: The exporter should work via socket, not the external IP.
Reason - Exporters fetch huge amounts of data, which creates a parasitic monitoring load.
to reduce the effect - avoid transferring via a network.
```suggestion
                    "mysql.host": "unix(/var/run/mysqld/mysqld.sock)",
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/mysql_vm_helpers.py [hunk 7]
```diff
@@ -163,6 +188,31 @@ def wait_until_mysql_connection(self) -> None:
...
+        """Set up mysqld-exporter config options.
+
+        Raises
+            snap.SnapError: if an issue occurs during config setting or restart
+        """
+        cache = snap.SnapCache()
+        mysqld_exporter = cache[MYSQL_EXPORTER_SNAP_NAME]
+
+        try:
+            mysqld_exporter.set(
```
Comment (delgod) [inline src/mysql_vm_helpers.py:201]: I don't know how to make it happen with a snap, but the following collectors should be disabled to reduce the parasitic monitoring load.
```
collect.auto_increment.columns
collect.info_schema.tables
collect.info_schema.tablestats
collect.perf_schema.indexiowaits
collect.perf_schema.tableiowaits
collect.perf_schema.tablelocks
collect.info_schema.userstats
collect.binlog_size
collect.info_schema.processlist
```

File: src/mysql_vm_helpers.py [hunk 8]
```diff
@@ -163,6 +188,41 @@ def wait_until_mysql_connection(self) -> None:
...
+                    "collect.info-schema.processlist": "false",
+                }
+            )
+            mysqld_exporter.restart()
+        except snap.SnapError as e:
+            logger.error(
+                "An exception occurred when setting configs for mysqld-exporter snap. Reason: %s"
+                % e.message
+            )
+            raise
```
Comment (shayancanonical) [inline src/mysql_vm_helpers.py:224]: minor nit: we have been raising custom errors for methods in this file so we can distinguish in the charm file where the exception originated and have conditional logic based on the method's exception

**Review by shayancanonical — APPROVED** [review]

**Comment by taurus-forever** [comment]
@zmraul JFYI, mysql-operator has [migrated](https://github.com/canonical/mysql-operator/pull/128) to the fat-snap, feel free to finish and merge this PR.
Thank you!

File: src/mysql_vm_helpers.py (hunk 8 above)
Comment (paulomach) [reply src/mysql_vm_helpers.py]: 8aba997

**Comment by paulomach** [comment]
Note to self:
*TODO*: add log files

**Review by delgod — APPROVED** [review]

**Pull request merged**